### DISCUSSION TITLE
Help needed getting 2nd HW I2C working on new ATMega328PB

### DISCUSSION THREAD

**Discussion by Balze2** [olikraus/u8g2#1442] [Q&A]
Hi all,

first of all:
@olikraus : **THANK YOU VERY MUCH!
Thanks for creating and maintaining this library!**

Sad but true I'm not a great C++ programmer like you.

I try to get your library running with my setup in HW I2C mode.
I have a ATMega328PB. This new (Atmel) Microchip device is the successor of the popular ATMega328P (which is on the Arduino Nano for example) and is pin compatible with it. Except 2 pins that have been Vcc and GND and are now two digital IO pins with an additional I2C bus.
My display is a 0,96" OLED display with a SSD1306 controller. It is connected to this 2nd I2C bus.

I get it working with SW I2C on pin 23 (SDA) and pin 24 (SCL) using this constructor.
U8G2_SSD1306_128X64_NONAME_F_SW_I2C u8g2(U8G2_R0, /* clock=*/ 24, /* data=*/ 23, /* reset=*/ U8X8_PIN_NONE);

Thus my hardware setup is okay.

But it does not work with the constructor for the second I2C, like this:
U8G2_SSD1306_128X64_NONAME_F_2ND_HW_I2C u8g2(U8G2_R0, /* reset=*/ U8X8_PIN_NONE);

I would appreciate any advice on how get it running with HW I2C.

Thank you very much in advance!

Regards

Balze2

**Comment by Balze2** [comment]
Some additional informations:

All the above mentioned is done with the Arduino IDE (1.8.13)
To be able to use the 328PB Microcontroller the "MCUdude/MiniCore" is used.
https://github.com/MCUdude/MiniCore

The core includes wire.h and wire1.h (*.cpp as well)

The arduino log has this line (single line, added \r\n for better reading):
```
"C:\\Users\\Balze2\\AppData\\Local\\Arduino15\\packages\\arduino\\tools\\avr-gcc\\7.3.0-atmel3.6.1-arduino7/bin/avr-g++" -c -g -Os -w
 -std=gnu++11 -fno-exceptions -ffunction-sections -fdata-sections -fno-threadsafe-statics -w -x c++ -E -CC -mmcu=atmega328pb -DF_CPU=20000000L
 -DARDUINO=10813 -DARDUINO_AVR_ATmega328 -DARDUINO_ARCH_AVR
 "-IC:\\Users\\Balze\\AppData\\Local\\Arduino15\\packages\\MiniCore\\hardware\\avr\\2.1.1\\cores\\MCUdude_corefiles"
 "-IC:\\Users\\Balze\\AppData\\Local\\Arduino15\\packages\\MiniCore\\hardware\\avr\\2.1.1\\variants\\pb-variant"
 "-IC:\\Users\\Balze\\Documents\\Arduino\\libraries\\U8g2\\src"
 "-IC:\\Users\\Balze\\AppData\\Local\\Arduino15\\packages\\MiniCore\\hardware\\avr\\2.1.1\\libraries\\SPI\\src"
 "-IC:\\Users\\Balze\\AppData\\Local\\Arduino15\\packages\\MiniCore\\hardware\\avr\\2.1.1\\libraries\\Wire\\src"
 "C:\\Users\\Balze\\AppData\\Local\\Temp\\arduino_build_395362\\sketch\\GraphicsTest.ino.cpp" -o nul -DARDUINO_LIB_DISCOVERY_PHASE
``` 
I'm wondering that wire is included and compiled and not wire1

Any help is greatly appreciated.

Regards

Balze2

**Comment by Balze2** [comment]
Update:

I tweaked the files for the 1st I2C port to use SDA1 and SCL1 and the TWI1 registers. ->  WIRE-files (wire.h; wire.cpp; twi.h and twi.c)
With this changes I used this constructor:
U8G2_SSD1306_128X64_NONAME_F_HW_I2C u8g2(U8G2_R0, /* reset=*/ U8X8_PIN_NONE);

Et voilà. It's working.

I kind of rerouted the 1st I2C to the pins of the 2nd one.
But this does not help a lot. Because I need both the I2C ports. 

Thus: Any help is greatly appreciated.

Regards

Balze2

**Reply by olikraus** [reply]
I assume U8G2_SSD1306_128X64_NONAME_F_2ND_HW_I2C is not supported. 
Actually u8g2 depends on proper setup of the i2c interface count, which is missing here, i guess.
More specific: Arduino IDE needs to specify WIRE_INTERFACES_COUNT to be 2 instead of 1.

**Reply by Balze2** [reply]
Thank you Oli for your reply.

OK, I understand. So I'll get in contact with the developer of he MiniCore and ask him to add/change this.
In fact the "wire" part is different to the Arduino Zero for example.

Best regards

Balze

**Comment by Balze2** [comment]
Thank you Oli for your hint.

I added 
`#define WIRE_INTERFACES_COUNT  2`
to the pins_arduino.h in the variants folder of the MiniCore.
And added 
`#include <wire1.h>`
to your "hello world" Example

Now it is working with the constructor:
`U8G2_SSD1306_128X64_NONAME_F_2ND_HW_I2C u8g2(U8G2_R0, /* reset=*/ U8X8_PIN_NONE);`

Thank you very much for your support!

Best regards

Balze

**Reply by Balze2** [reply]
it is not necessary to include wire1.h.
this is obviously done by dependencies.

**Reply by 13507086298** [reply]
> 
I added according to your steps 
`#define WIRE_INTERFACES_COUNT 2` 
to pins_arduino.h file document in the variants folder of the MiniCore.
and used 
`U8G2_SSD1306_128X64_NONAME_F_2ND_HW_I2C u8g2(U8G2_R0, /* reset=*/ U8X8_PIN_NONE);`

But he doesn't show, it doesn't seem to work .

Can you give a more detailed how-to? 

Hope to get your reply, thanks

**Comment by Balze2** [comment]
Did you verify the electrical connection to the second I2C?
If so, did you try my "reroute" method mentioned above?
Did you ever had the display show anything? (On another port or with another lib?)

**Comment by 13507086298** [comment]
yes i used this constructor:
`U8G2_SSD1306_128X64_NONAME_F_SW_I2C u8g2(U8G2_R0, /* clock=*/ 24, /* data=*/ 23, /* reset=*/ U8X8_PIN_NONE); `
This can be displayed normally, but using 
`U8G2_SSD1306_128X64_NONAME_F_2ND_HW_I2C u8g2(U8G2_R0, /* reset=*/ U8X8_PIN_NONE);`
and
`U8G2_SSD1306_128X64_NONAME_F_HW_I2C u8g2(U8G2_R0, /* reset=*/ U8X8_PIN_NONE, /* clock=*/ 24, /* data=*/ 23);`
nothing shows up .(I have added `#define WIRE_INTERFACES_COUNT 2` into the pins_arduino.h file )

Thank you very much for your help.

**Comment by Balze2** [comment]
Now I see.
If you use a SW I2C setup on Pins 23 and 24 it works.
This means you have the display connected to pins 23 and 24.
There is no HW I2C on these pins on a 328PB.
And you can't remap I2C to these pins.
Look for the HW pins in datasheet and reconnect your display.
I2C_1: pins 27 and 28
I2C_2: pins 3 and 6

Good luck

Balze2

**Comment by 13507086298** [comment]
Uh, sorry. It should be that I didn't make it clear, the pins 23 and 24 used in my constructor correspond to hardware pins 3 and 6.

The electrical connection has been checked many times, and I can confirm that the I2C_1 and I2C_2 hardware connections are correct (I have 2 I2C devices), and I have checked SCL1 (pins 24, hardware pin 6) and SDA1 (pins 23, hardware pin 3) and tested the way you rerouted, but it still doesn't work. So I hope you can give some more detailed steps.

Thank you for your help

maxpa

**Comment by Balze2** [comment]
Are you aware, that you need pull up resistors for SDA and SCL lines if you use HW I2C??

**Comment by 13507086298** [comment]
Thank you for your patient reply, that's the problem
